cloudnew/ec2_util.go

```diff
@@ -140,3 +155,334 @@ func timeTilNextHourlyPayment(host *host.Host) time.Duration {
+func fetchEBSPricing() (map[string]float64, error) {
+	// there is no true EBS pricing API, so we have to wrangle it from EC2's frontend
+	endpoint := "http://a0.awsstatic.com/pricing/1/ebs/pricing-ebs.js"
+	grip.Debugln("Loading EBS pricing from", endpoint)
+
+	client := util.GetHttpClient()
+	defer util.PutHttpClient(client)
+
+	resp, err := client.Get(endpoint)
+	if resp != nil {
+		defer resp.Body.Close()
+	}
+	if err != nil {
+		return nil, errors.Wrapf(err, "fetching %s", endpoint)
+	}
+	data, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errors.Wrap(err, "reading response body")
+	}
+	matches := ebsRegex.FindSubmatch(data)
+	if len(matches) < 2 {
+		return nil, errors.Errorf("could not find price JSON in response from %v", endpoint)
+	}
+	// define a one-off type for storing results from the price JSON
+	prices := struct {
+		Config struct {
+			Regions []struct {
+				Region string
+				Types  []struct {
+					Name   string
+					Values []struct {
+						Prices struct {
+							USD string
+						}
+					}
+				}
+			}
+		}
+	}{}
+	err = json.Unmarshal(matches[1], &prices)
+	if err != nil {
+		return nil, errors.Wrap(err, "parsing price JSON")
+	}
+	fmt.Printf("%+v\n", prices)
+
+	pricePerRegion := map[string]float64{}
+	for _, r := range prices.Config.Regions {
+		for _, t := range r.Types {
+			// only cache "general purpose" pricing for now
+			if strings.Contains(t.Name, "ebsGPSSD") {
+				if len(t.Values) == 0 {
+					continue
+				}
+				price, err := strconv.ParseFloat(t.Values[0].Prices.USD, 64)
+				if err != nil {
+					continue
+				}
+				pricePerRegion[r.Region] = price
+			}
+		}
+	}
+	// one final sanity check that we actually pulled information, which will alert
+	// us if, say, Amazon changes the structure of their JSON
+	if len(pricePerRegion) == 0 {
+		return nil, errors.Errorf("unable to parse prices from %v", endpoint)
+	}
+	return pricePerRegion, nil
+}
+
+// blockDeviceCosts returns the total price of a slice of BlockDevices over the given duration
+// by using the EC2 API.
+// ebsCost returns the cost of running an EBS block device for an amount of time in a given size and region.
+// EBS bills are charged in "GB/Month" units. We consider a month to be 30 days.
+func ebsCost(pf ebsPriceFetcher, region string, size int64, duration time.Duration) (float64, error) {
+	prices, err := pf.FetchEBSPrices()
+	if err != nil {
+		return 0.0, err
+	}
+	price, ok := prices[region]
+	if !ok {
+		return 0.0, errors.Errorf("no EBS price for region '%v'", region)
+	}
+	// price = GB * % of month *
+	month := (time.Hour * 24 * 30)
+
+	return float64(size) * (float64(duration) / float64(month)) * price, nil
+
+}
+
+// onDemandPriceFetcher is an interface for fetching the hourly price of a given
+// os/instance/region combination.
+type onDemandPriceFetcher interface {
+	FetchPrice(os osType, instance, region string) (float64, error)
```

Comment:
does it make sense to keep these interfaces seperate for different kinds of providers? should we just merge them into a single interface?